src/api/app/models/event_subscription/generate_hash_for_subscriber.rb

```diff
@@ -0,0 +1,72 @@
+class EventSubscription
+  class GenerateHashForSubscriber
+    attr_reader :subscriber
+
+    def initialize(subscriber)
+      @subscriber = subscriber
+    end
+
+    # Generate a hash of subscriptions for the user grouped by the subscription's event class
+    def query
+      event_classes = Event::Base.notification_events
+      subscriptions_by_event_class = {}
+
+      event_classes.each do |event_class|
+        event_class.receiver_roles.each do |role|
+          subscription = find_subscription_for_event_class_and_role(event_class, role)
+
+          subscriptions_by_event_class[event_class] ||= []
+          subscriptions_by_event_class[event_class] << subscription
+        end
```

Comment:
It would change the behaviour a little bit. When event_class.receiver_roles is empty, subscriptions_by_event_class[event_class] would now be an empty Array instead of nil.